### PR TITLE
Add text derivative

### PR DESCRIPTION
**Pull request by shevTessella**
Add getTextDerivative function. 
call function by $("#my-plate-layout").plateLayOut("getTextDerivative");

### CODE REVIEW

File: app/assets/javascripts/plate-map/add-data-on-change.js [hunk 1]
```diff
@@ -167,7 +167,7 @@ var plateLayOutWidget = plateLayOutWidget || {};
           "selectedAreas": selectedAreas,
           "focalWell": focalWell,
           "colorToLoc": colorLocMap,
-          "requiredField": requiredField
+          "requiredField": requiredField,
```
Comment (gorj-tessella) [inline app/assets/javascripts/plate-map/add-data-on-change.js]: A comma here shouldn't affect newer browsers but can cause issues with some javascript engines.

File: app/assets/javascripts/plate-map/plate-layout.js [hunk 2]
```diff
@@ -133,5 +133,28 @@ $.widget("DNA.plateLayOut", {
...
+    var textDerivative = {};
+    var fieldMap = this.fieldMap;
+    for (var idx in derivative){
+      var textValWell = {};
+      var textFieldIdWell = {};
+      var curWellData = derivative[idx].wellData;
+      for (var fieldId in curWellData){
+        var field = this.fieldMap[fieldId];
+        textFieldIdWell[field.name] = field.parseText(curWellData[fieldId]);
+        textValWell[fieldId] = field.parseText(curWellData[fieldId]);
```
Comment (gorj-tessella) [inline app/assets/javascripts/plate-map/plate-layout.js]: Should parseText only once for performance

File: app/assets/javascripts/plate-map/create-field.js [hunk 3]
```diff
@@ -1090,6 +1128,39 @@ var plateLayOutWidget = plateLayOutWidget || {};
           }
           that.fieldWarningMsg(field, warningText, mainFieldWarning);
         };
+
+        field.parseMainFieldVal = function(val) {
+          var optMap = field.data.options;
```
Comment (gorj-tessella) [inline app/assets/javascripts/plate-map/create-field.js]: You are iterating over field.data.options (which is an array) as if it is an object. Use: 

```
for (var idx = 0; idx < optMap.length; idx++) {
```

Comment (shevTessella) [reply app/assets/javascripts/plate-map/create-field.js]: Resolved

**Comment by shevTessella** [comment]
Resolved all of the comments and addded getWellsDifferences method

File: app/assets/javascripts/plate-map/fabric-events.js [hunk 4]
```diff
@@ -323,72 +323,102 @@ var plateLayOutWidget = plateLayOutWidget || {};
...
+            var diffWellVal = {};
+            var curWellData = wells[wellIdx];
+            for (var fieldId in curWellData) {
+              var commonVal = commonWell[fieldId];
+              var curVal = curWellData[fieldId];
+              var newVal = null;
+              if (Array.isArray(curVal)) {
+                // get uncommonVal
+                newVal = [];
+                if (curVal[0] && typeof(curVal[0] === "object")){
```
Comment (gorj-tessella) [inline app/assets/javascripts/plate-map/fabric-events.js]: Shouldn't each entry in curVal be checked, not just the first?

Comment (shevTessella) [reply app/assets/javascripts/plate-map/fabric-events.js]: The way it works here is that if the multiplex field is empty, then the multiplex field in this well is an empty array. If the multiplex field has value, then the field will be [{subfield:val, subfield2:val2}, {subfield:val}]. I think checking the first entry is sufficient to be identified as a multiplexfield.

**Pull request merged**